syntax = "proto3";

package itinerary;

service ItineraryService {
  rpc GetItinerary(Empty) returns (ItineraryList);
  rpc UpdateItinerary(UpdateRequest) returns (UpdateResponse);
  rpc StreamItineraryChanges(Empty) returns (stream ItineraryList);

  // Replication
  rpc ReplicateItinerary(UpdateRequest) returns (Empty);

  // Heartbeats and leader detection
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
}

message ItineraryItem {
  int32 id = 1;
  string name = 2;
  int32 number = 3;
}

message ItineraryList {
  repeated ItineraryItem items = 1;
}

message UpdateRequest {
  int32 itinerary_id = 1;
  int32 quantity_change = 2;
}

message UpdateResponse {
  bool success = 1;
}

message HeartbeatRequest {
  string host = 1;
  int32 port = 2;
}

message HeartbeatResponse {
  bool is_leader = 1;
}

message Empty {}
